### PR TITLE
【Hackathon 6th No.37】GraphCastNet 代码迁移至 PaddleScience

### PR DESCRIPTION
**Pull request by MayYouBeProsperous**
### PR types

Others

### PR changes

APIs, Docs

### Describe

复现运行结果
image

### CODE REVIEW

**Comment by MayYouBeProsperous** [comment]
麻烦制作数据集下载链接：https://aistudio.baidu.com/datasetdetail/252766

**Comment by lijialin03** [comment]
> 麻烦制作数据集下载链接：https://aistudio.baidu.com/datasetdetail/252766

数据集下载链接分别为：
[dataset.zip](https://paddle-org.bj.bcebos.com/paddlescience/datasets/graphcast/dataset.zip)
[dataset.zip](https://paddle-org.bj.bcebos.com/paddlescience/datasets/graphcast/dataset.zip)
[dataset.zip](https://paddle-org.bj.bcebos.com/paddlescience/datasets/graphcast/dataset.zip)
[dataset.zip](https://paddle-org.bj.bcebos.com/paddlescience/datasets/graphcast/dataset.zip)

**Comment by lijialin03** [comment]
我看标题中的WIP去掉了，代码现在是可以review的状态嘛

**Comment by MayYouBeProsperous** [comment]
是的，可以 review 了。
CI 报错好像是注释的括号匹配问题，应该怎么解决呢
image
image

**Comment by lijialin03** [comment]
> 是的，可以 review 了。 CI 报错好像是注释的括号匹配问题，应该怎么解决呢 image image

好嘞，这个先不管吧，等review完再说

**Review by lijialin03 — APPROVED** [review]
调用jointContribution/graphcast/run中的compare
输入运行结果后得到对齐结果为：
```
10m_u_component_of_wind diff is -2.356016193110792e-05
10m_v_component_of_wind diff is 7.523250922509292e-06
2m_temperature diff is 2.4777781459127155e-05
geopotential diff is -0.000539646573504715
mean_sea_level_pressure diff is -0.0005250472340893714
specific_humidity diff is 6.035347363899133e-10
temperature diff is 8.158555083697042e-07
total_precipitation_6hr diff is 1.8224237754237203e-09
u_component_of_wind diff is 1.7749217200032017e-06
v_component_of_wind diff is 7.323719328271271e-06
vertical_velocity diff is -2.375314757111974e-07
All diff is -8.92277301779838e-05
```
参考[https://github.com/PaddlePaddle/PaddleScience/pull/699](https://github.com/PaddlePaddle/PaddleScience/pull/699)，认为可以对齐

File: examples/graphcast/graphcast.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    # initialize solver
+    solver = ppsci.solver.Solver(
+        model,
+        output_dir=cfg.output_dir,
+        log_freq=cfg.log_freq,
+        seed=cfg.seed,
+        validator=validator,
+        pretrained_model_path=cfg.EVAL.pretrained_model_path,
+        eval_with_no_grad=cfg.EVAL.eval_with_no_grad,
```
Comment (HydrogenSulfate) [inline examples/graphcast/graphcast.py:108]:
```suggestion
        model,
        validator=validator,
        cfg=cfg,
```

File: ppsci/arch/graphcast.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+            mesh_node_emb_dim (int): Dimension of emdding mesh nodes.
+            mesh_edge_dim (int): Dimension of mesh edges.
+            mesh_edge_emb_dim (int): Dimension of emdding mesh edges.
+            grid2mesh_edge_dim (int): Dimension of mesh edges in Grid2Mesh GNN.
+            grid2mesh_edge_emb_dim (int): Dimension of emdding mesh edges in Grid2Mesh GNN.
+            mesh2grid_edge_dim (int): Dimension of mesh edges in Mesh2Grid GNN.
+            mesh2grid_edge_emb_dim (int): Dimension of emdding mesh edges in Mesh2Grid GNN.
+            gnn_msg_steps (int): Step of gnn messages.
+            node_output_dim (int): Dimension of output nodes.
+        """
```
Comment (HydrogenSulfate) [inline ppsci/arch/graphcast.py:430]: class的docstring放到class名称下方，不放到__init__下面

File: ppsci/data/dataset/atmospheric_dataset.py [hunk 3]
```diff
@@ -0,0 +1,1779 @@
...
+            grid_num_nodes (int, optional): Number of grid nodes. Defaults to None.
+            mesh_num_edges (int, optional): Number of mesh edges. Defaults to None.
+            grid2mesh_num_edges (int, optional): Number of edges in Grid2Mesh GNN. Defaults to None.
+            mesh2grid_num_edges (int, optional): Number of edges in Mesh2Grid GNN. Defaults to None.
+            grid_node_feat (np.array, optional): Feature of grid nodes. Defaults to None.
+            mesh_node_feat (np.array, optional): Feature of mehs nodes. Defaults to None.
+            mesh_edge_feat (np.array, optional): Feature of mesh edges. Defaults to None.
+            grid2mesh_edge_feat (np.array, optional): Feature of edges in Grid2Mesh GNN. Defaults to None.
+            mesh2grid_edge_feat (np.array, optional): Feature of edges in Mesh2Grid GNN. Defaults to None.
+        """
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/atmospheric_dataset.py:870]: 同上，移动到class名称的下方，不放在__init__里

File: ppsci/data/dataset/atmospheric_dataset.py [hunk 4]
```diff
@@ -0,0 +1,1779 @@
...
+AVG_SEC_PER_YEAR = SEC_PER_DAY * _AVG_DAY_PER_YEAR
+
+DAY_PROGRESS = "day_progress"
+YEAR_PROGRESS = "year_progress"
+
+
+def stacked_to_dataset(
+    stacked_array: "xarray.Variable",
+    template_dataset: "xarray.Dataset",
+    preserved_dims: typing.Tuple[str, ...] = ("batch", "lat", "lon"),
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/atmospheric_dataset.py:204]: typing.XXX建议改为XXX

File: ppsci/data/dataset/atmospheric_dataset.py [hunk 5]
```diff
@@ -0,0 +1,1779 @@
...
+        # to xarray `Dataset` (batch, one time step, lat, lon, level, multiple vars)
+        return stacked_to_dataset(grid_xarray.variable, targets_template)
+
+
+class TriangularMesh(typing.NamedTuple):
+    vertices: np.ndarray
+    faces: np.ndarray
+
+
+def merge_meshes(mesh_list: typing.Sequence[TriangularMesh]) -> TriangularMesh:
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/atmospheric_dataset.py:1149]: 同上

File: ppsci/data/dataset/atmospheric_dataset.py [hunk 6]
```diff
@@ -0,0 +1,1779 @@
...
+    vertices = np.dot(vertices, rotation_matrix)
+
+    return TriangularMesh(
+        vertices=vertices.astype(np.float32), faces=np.array(faces, dtype=np.int32)
+    )
+
+
+def get_hierarchy_of_triangular_meshes_for_sphere(
+    splits: int,
+) -> typing.List[TriangularMesh]:
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/atmospheric_dataset.py:1213]: 同上

File: ppsci/data/dataset/atmospheric_dataset.py [hunk 7]
```diff
@@ -0,0 +1,1779 @@
...
+                [ind12, ind23, ind31],  # 4
+            ]
+        )
+    return TriangularMesh(
+        vertices=new_vertices_builder.get_all_vertices(),
+        faces=np.array(new_faces, dtype=np.int32),
+    )
+
+
+class _ChildVerticesBuilder(object):
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/atmospheric_dataset.py:1247]: python3不需要显式继承object
```suggestion
class _ChildVerticesBuilder:
```

File: ppsci/data/dataset/spherical_swe_dataset.py
Comment (HydrogenSulfate) [inline ppsci/data/dataset/spherical_swe_dataset.py:1]: 这类位置不需要加换行，其余文件同理，不需要添加换行

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
辛苦提交PR，代码内容修改完毕之后，可以同步更新一下文档的代码块定位位置

File: docs/zh/examples/graphcast.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
+# GraphCast
+
+=== "模型评估命令"
+    ``` sh
+
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply docs/zh/examples/graphcast.md]: Done.

File: docs/zh/examples/graphcast.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    # curl https://paddle-org.bj.bcebos.com/paddlescience/datasets/graphcast/stats.zip -o stats.zip
+    # curl https://paddle-org.bj.bcebos.com/paddlescience/datasets/graphcast/graphcast-jax2paddle.csv --create-dirs -o ./data/graphcast-jax2paddle.csv
+
+    unzip -q dataset.zip -d data/
+    unzip -q dataset-step12.zip -d data/
+    unzip -q params.zip -d data/
+    unzip -q stats.zip -d data/
+    unzip -q template_graph.zip -d data/
+
+    python graphcast.py
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply docs/zh/examples/graphcast.md]: Done.

File: examples/graphcast/conf/graphcast_small.yaml [hunk 10]
```diff
@@ -0,0 +1,63 @@
+hydra:
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply examples/graphcast/conf/graphcast_small.yaml:11]: Done.

File: examples/graphcast/conf/graphcast_small.yaml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    chdir: false # keep current working direcotry unchaned
+    config:
+      override_dirname:
+        exclude_keys:
+          - TRAIN.checkpoint_path
+          - TRAIN.pretrained_model_path
+          - EVAL.pretrained_model_path
+          - mode
+          - output_dir
+          - log_freq
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply examples/graphcast/conf/graphcast_small.yaml]: Done.

File: examples/graphcast/conf/graphcast_small.yaml [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  mesh2grid_edge_dim: 4
+  mesh2grid_edge_emb_dim: 512
+  grid2mesh_edge_dim: 4
+  grid2mesh_edge_emb_dim: 512
+  gnn_msg_steps: 16
+  node_output_dim: 83
+
+EVAL:
+  batch_size: 1
+  pretrained_model_path: "data/params/GraphCast_small---ERA5-1979-2015---resolution-1.0---pressure-levels-13---mesh-2to5---precipitation-input-and-output.pdparams"
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply examples/graphcast/conf/graphcast_small.yaml]: Done.

File: examples/graphcast/graphcast.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+            "label_keys": ("label",),
+            **cfg.DATA,
+        },
+        "batch_size": cfg.EVAL.batch_size,
+        "sampler": {
+            "name": "BatchSampler",
+            "drop_last": False,
+            "shuffle": False,
+        },
+        "num_workers": 1,
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply examples/graphcast/graphcast.py]: Done.

File: examples/graphcast/graphcast.py (hunk 1 above)
Comment (MayYouBeProsperous) [reply examples/graphcast/graphcast.py]: Done.

File: ppsci/data/__init__.py [hunk 14]
```diff
@@ -135,6 +134,20 @@ def build_dataloader(_dataset, cfg):
...
+            collate_fn = batch_transform.default_collate_fn
+        dataloader_ = io.DataLoader(
+            dataset=_dataset,
+            places=device.get_device(),
+            batch_sampler=batch_sampler,
+            collate_fn=collate_fn,
+            num_workers=cfg.get("num_workers", _DEFAULT_NUM_WORKERS),
+            use_shared_memory=cfg.get("use_shared_memory", False),
+            worker_init_fn=init_fn,
+        )
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply ppsci/data/__init__.py:150]: 区别是要设置 collate_fn = batch_transform.default_collate_fn

File: ppsci/data/dataset/spherical_swe_dataset.py
Comment (MayYouBeProsperous) [reply ppsci/data/dataset/spherical_swe_dataset.py:1]: 这里是删除多余换行的

File: ppsci/arch/graphcast.py [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+        self,
+        grid_node_num: int,
+        grid_node_emb_dim: int,
+        mesh_node_num: int,
+        mesh_node_emb_dim: int,
+        mesh_edge_emb_dim: int,
+        grid2mesh_edge_emb_dim: int,
+        mesh2grid_edge_emb_dim: int,
+        src_type="mesh",
+        dst_type="mesh",
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply ppsci/arch/graphcast.py]: Done.

File: ppsci/arch/graphcast.py [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+        graph.grid_node_feat = grid_node_emb
+        graph.mesh_node_feat = mesh_node_emb
+        graph.mesh_edge_feat = mesh_edge_emb
+        graph.grid2mesh_edge_feat = grid2mesh_edge_emb
+        graph.mesh2grid_edge_feat = mesh2grid_edge_emb
+
+        return graph
+
+
+class GraphCastGrid2Mesh(paddle.nn.Layer):
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply ppsci/arch/graphcast.py]: Done.

File: ppsci/arch/graphcast.py [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+            GraphCastMLP(grid_node_emb_dim, grid_node_emb_dim)
+        )
+
+    def forward(self, graph: "atmospheric_dataset.GraphGridMesh"):
+        graph = self.grid2mesh_gnn(graph)
+        graph.grid_node_feat = self.grid_node_layer(graph.grid_node_feat)
+        return graph
+
+
+class GraphCastMesh2Grid(paddle.nn.Layer):
```
[the remark this answers is not in the document]
Comment (MayYouBeProsperous) [reply ppsci/arch/graphcast.py]: Done.

File: ppsci/arch/graphcast.py (hunk 2 above)
Comment (MayYouBeProsperous) [reply ppsci/arch/graphcast.py]: Done.

File: ppsci/data/dataset/atmospheric_dataset.py (hunk 4 above)
Comment (MayYouBeProsperous) [reply ppsci/data/dataset/atmospheric_dataset.py]: 好的，已修改

File: ppsci/data/dataset/atmospheric_dataset.py (hunk 3 above)
Comment (MayYouBeProsperous) [reply ppsci/data/dataset/atmospheric_dataset.py]: Done.

File: ppsci/data/dataset/atmospheric_dataset.py (hunk 5 above)
Comment (MayYouBeProsperous) [reply ppsci/data/dataset/atmospheric_dataset.py]: Done.

File: ppsci/data/dataset/atmospheric_dataset.py (hunk 6 above)
Comment (MayYouBeProsperous) [reply ppsci/data/dataset/atmospheric_dataset.py]: Done.

File: ppsci/data/dataset/atmospheric_dataset.py (hunk 7 above)
Comment (MayYouBeProsperous) [reply ppsci/data/dataset/atmospheric_dataset.py]: Done.

**Review by HydrogenSulfate — COMMENTED** [review]
LGTM

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
